Threadsafe futures

ibaryshnikov opened the pull request:
Closes #1379 

Changes:
- Added extern function `wait_async`, which does the same as `Atomics::waitAsync`
- Removed `impl Notify for Package`, also removed `Send + Sync` markers for `Package`
- Added `Waker` struct instead, which implements `Notify`
- Used `Waker` instead of `Package` to notify the executor

ibaryshnikov commented on crates/js-sys/src/lib.rs:
```diff
@@ -478,6 +478,9 @@ extern "C" {
     pub fn slice_with_end(this: &SharedArrayBuffer, begin: u32, end: u32) -> SharedArrayBuffer;
 }
 
+unsafe impl Send for SharedArrayBuffer {}
```
@alexcrichton is it fine to mark `SharedArrayBuffer` as `Send + Sync` manually? I tried to remove `#[wasm_bindgen(extends = Object)]`, but the compiler still thought that `SharedArrayBuffer` had a `PhantomData` from `JsValue`

alexcrichton replied:
Unfortunately this can't be done because the `JsValue` underlying can't actually be sent between threads (inherently). The `SharedArrayBuffer` value that you'll be notifying through will need to be `wasm_bindgen::memory()` I believe

ibaryshnikov replied:
It's an interesting point. So I can remove `unsafe impl` from here?
I have some more questions about implementation.
- WebAssembly memory could be just an ArrayBuffer, not the shared one. Is it fine?
- In docs there is `pub fn memory() -> JsValue`, which refers to the whole linear memory. Do I need to use `std::alloc` here to obtain a proper memory location?

alexcrichton replied:
Oh sorry forgot to follow-up but yes it should be safe to remove these.

ibaryshnikov commented on crates/futures/src/lib.rs:
```diff
@@ -339,47 +352,48 @@ fn _future_to_promise(future: Box<Future<Item = JsValue, Error = JsValue>>) -> P
-        }
-    }
+    //    impl Notify for Package {
+    //        fn notify(&self, _id: usize) {
```
@alexcrichton I added a `Waker` struct with a simple `Notify` implementation, but I don't know yet where to put all the logic from `impl Notify for Package`. Can you suggest a good place for it?

alexcrichton commented on crates/futures/src/polyfill.js:
```diff
@@ -0,0 +1,93 @@
+const helperCode = `
```
Could this file have a comment about where it comes from?

ibaryshnikov replied:
Oh, thanks!

ibaryshnikov commented:
Does it mean that the state with `Polling | Notified | Waiting` will be removed, and there will be only `true` or `false`?

alexcrichton commented:
Oh what I gisted was just a straw man, it may be that two bits are still needed rather than one for information exchange.

ibaryshnikov commented:
@alexcrichton the pr is not tested yet, I pushed changes to check the design.
There are several places to check: first, I removed `impl Notify for Package`. Then, instead of `&me`, there is `&me.waker` passed to `poll_future_notify`.
And I moved code from `impl Notify for Package` to `fn poll_again`
Inside `fn poll_again` `Promise::resolve` was replaced with `wait_async`
Did I get your idea right?

alexcrichton commented on crates/futures/src/lib.rs:
```diff
@@ -289,10 +296,68 @@ fn _future_to_promise(future: Box<Future<Item = JsValue, Error = JsValue>>) -> P
-    unsafe impl Send for Package {}
-    unsafe impl Sync for Package {}
+    struct Waker {
+        buffer: SharedArrayBuffer,
```
This I don't think will store `SharedArrayBuffer` but rather use `wasm_bindgen::memory().buffer()` as the `SharedArrayBuffer` accessor

ibaryshnikov replied:
oh, fine! but how do I allocate it?

ibaryshnikov commented:
@alexcrichton I think I hit the wall here: https://github.com/rustwasm/wasm-bindgen/issues/1525
There's a need to either make js snippets work with node target or replace the snippet with polyfill with the wasm equivalent. What do you think is better?

alexcrichton commented:
Yeah it may be some time before snippets work on Node, so it's probably best to not block this on that. Perhaps the polyfill could be implemented in Rust though?

alexcrichton commented on crates/futures/src/polyfill.rs:
```diff
@@ -0,0 +1,209 @@
+    console_log!("polyfill, general case");
+
+    Ok(Promise::new(
+        &mut Box::new(move |resolve: Function, reject: Function| {
```
I think this box may be able to be removed?

alexcrichton commented on crates/futures/src/polyfill.js:
```diff
@@ -0,0 +1,130 @@
+/*
+ * The polyfill was kindly borrowed from https://github.com/tc39/proposal-atomics-wait-async
```
I think this file may no longer be necessary?

alexcrichton commented on crates/futures/src/polyfill.rs:
```diff
@@ -0,0 +1,209 @@
+                .borrow()
+                .set_onmessage(Some(onmessage_callback.as_ref().unchecked_ref()));
+
+            onmessage_callback.forget();
```
I think we'll want to avoid `forget` here since it'll leak memory quickly, perhaps the `Closure::once` constructor could be used?

ibaryshnikov replied:
`Closure::once` still produces `Error: closure invoked recursively or destroyed already` if I remove `forget`. May be `Closure::once_into_js` instead?

alexcrichton commented on crates/futures/src/polyfill.rs:
```diff
@@ -0,0 +1,209 @@
+";
+
+thread_local! {
+    static HELPERS: RefCell<Vec<Rc<RefCell<Worker>>>> = RefCell::new(vec![]);
```
Is the `RefCell` inside the `Rc` needed here to mutate the `Worker`?

alexcrichton commented on crates/futures/src/polyfill.rs:
```diff
@@ -0,0 +1,209 @@
+}
+
+pub fn wait_async(indexed_array: Int32Array, index: u32, value: i32) -> Result<Promise, JsValue> {
+    let timeout = 0.0;
```
I think this default timeout may be causing a spin loop?

ibaryshnikov replied:
Wow, changing default timeout to 10.0 helped! In spec it's a number of milliseconds. But I'm not sure which number to chose as default. What do you think?

alexcrichton commented on examples/raytrace-parallel/Cargo.toml:
```diff
@@ -13,7 +13,7 @@ futures = "0.1"
 raytracer = { git = 'https://github.com/alexcrichton/raytracer', branch = 'update-deps' }
 wasm-bindgen = { version = "0.2.45", features = ['serde-serialize'] }
-wasm-bindgen-futures = "0.3.22"
+wasm-bindgen-futures = { path = '../../crates/futures', version = "0.3.22" }
```
If this `path` was necessary, mind avoiding this change and instead updating `/Cargo.toml` and the `[patch]` section?

alexcrichton commented on crates/futures/src/lib.rs:
```diff
@@ -110,14 +110,35 @@ pub mod futures_0_3;
+#[cfg(target_feature = "atomics")]
+use js_sys::{Atomics, Int32Array, SharedArrayBuffer, WebAssembly};
 use wasm_bindgen::prelude::*;
+#[cfg(target_feature = "atomics")]
```
Organizationally could there perhaps be:

```
src/stable.rs // what lib.rs currently is
src/atomics.rs // what this file is currently
src/futures03.rs // implementation with futures 0.3
```

and eventually we can add `src/futures03_atomics.rs`, but for now we'd just go with what's here. I don't think we'll benefit too too much from shared types, but the shared pieces can always be in `src/lib.rs`

ibaryshnikov replied:
I moved current implementation to the `atomics.rs` file. Do I need to go farther and move contents of `lib.rs` to `stable.rs`?

alexcrichton commented on crates/futures/src/lib.rs:
```diff
@@ -110,14 +110,35 @@ pub mod futures_0_3;
+#[wasm_bindgen]
+extern "C" {
+    #[wasm_bindgen(js_namespace = console)]
+    fn log(s: &str);
```
Just a reminder that we'll want to remove these before landing.

alexcrichton commented on crates/futures/src/lib.rs:
```diff
@@ -289,9 +316,108 @@ fn _future_to_promise(future: Box<dyn Future<Item = JsValue, Error = JsValue>>)
+    #[cfg(target_feature = "atomics")]
+    fn poll_again(package: Arc<Package>, id: usize) {
+        console_log!("poll_again called");
+        let me = match package.notified.replace(State::Notified) {
```
This `notified` check will want to be the same as the `notified` field in `Waker` to ensure all the races align well

ibaryshnikov replied:
I don't get it yet. Adding the following check after the `match` statement fixes performance problems, but breaks the whole thing
```rust
if package.waker.notified.swap(true, Ordering::SeqCst) {
    return;
}
```

alexcrichton commented on crates/futures/src/lib.rs:
```diff
@@ -289,9 +316,108 @@ fn _future_to_promise(future: Box<dyn Future<Item = JsValue, Error = JsValue>>)
+        //
+        // Note that the `Rc`/`RefCell` trick here is basically to just
+        // ensure that our `Closure` gets cleaned up appropriately.
+        let promise = polyfill::wait_async(array, id as u32, 0)
```
I don't think that this is quite right because `id` is arbitrarily passed in. I think it's safe to effectively ignore the `id` in the trait method and everywhere else, and this should be a `waitAsync` on the memory location of `notified` in the `Waker` (which can be switched to `AtomicI32` to ensure it's a 32-bit integer)

alexcrichton commented on crates/futures/src/lib.rs:
```diff
@@ -289,9 +316,108 @@ fn _future_to_promise(future: Box<dyn Future<Item = JsValue, Error = JsValue>>)
+
+        let array_location = package.waker.array.as_ptr() as u32 / 4;
+        let array = Int32Array::new(&memory_buffer)
+            .subarray(array_location, array_location + package.waker.array.len() as u32);
```
I think it's safe to avoid the `subarray` here and just use the `Int32Array` with a larger offset that is the pointer that we're waiting on.

alexcrichton commented on crates/futures/src/lib.rs:
```diff
@@ -289,9 +316,108 @@ fn _future_to_promise(future: Box<dyn Future<Item = JsValue, Error = JsValue>>)
+        // ensure that our `Closure` gets cleaned up appropriately.
+        let promise = polyfill::wait_async(array, id as u32, 0)
+            .expect("Should create a Promise");
+        let slot = Rc::new(RefCell::new(None));
```
This dance may be a bit easier today with `Closure::once` perhaps?

alexcrichton commented on crates/futures/src/lib.rs:
```diff
@@ -289,9 +316,108 @@ fn _future_to_promise(future: Box<dyn Future<Item = JsValue, Error = JsValue>>)
+                let array = Int32Array::new(&memory_buffer)
+                    .subarray(array_location, array_location + self.array.len() as u32);
+
+                let _ = Atomics::notify(&array, id as u32);
```
I think this can be the [notify intrinsic in Rust](https://doc.rust-lang.org/core/arch/wasm32/fn.atomic_notify.html) perhaps?

alexcrichton commented on crates/futures/src/polyfill.rs:
```diff
@@ -0,0 +1,209 @@
+    });
+}
+
+pub fn wait_async(indexed_array: Int32Array, index: u32, value: i32) -> Result<Promise, JsValue> {
```
FWIW we can probably simplify this polyfill if it's written in Rust to simply take a pointer here instead of an `Int32Array` plus an `index`

ibaryshnikov replied:
So I can just remove `indexed_array` here, and our index will be the pointer? Since pointers in wasm32 are simply u32 numbers

ibaryshnikov commented on crates/futures/src/atomics.rs:
```diff
@@ -0,0 +1,351 @@
+            console_log!("Waker notify");
+            if !self.notified.swap(true, Ordering::SeqCst) {
+                console_log!("Waker, inside if");
+                let _ = unsafe { core::arch::wasm32::atomic_notify(&self.value as *const AtomicI32 as *mut i32, 0) };
```
@alexcrichton I have hard time trying to figure out how to use intrinsics here, may be you know a better way?

ibaryshnikov commented on crates/futures/src/atomics.rs:
```diff
@@ -0,0 +1,351 @@
+        Waiting(Arc<Package>),
+    }
+
+    struct Waker {
```
@alexcrichton so we can use only one field here, `notified`, and make it `AtomicI32`?

ibaryshnikov commented on crates/futures/src/lib.rs:
````diff
@@ -101,12 +101,22 @@
 //! }
 //! ```
 
+#![feature(stdsimd)]
````
@alexcrichton it's not correct to leave it here, but when I put it inside `atomics.rs`, the example fails to compile, with error mentioning this feature

alexcrichton replied:
Oh no this is correct yeah since the atomic intrinsics aren't stable yet, but you'll want to put this behind a `cfg_attr` to only compile it in when the `atomics` feature is enabled. That way this can continue to compile on stable.

ibaryshnikov commented on crates/futures/src/atomics.rs:
```diff
@@ -0,0 +1,350 @@
+            }
+        };
+
+        let value_location = &package.waker.value as *const AtomicI32 as u32 / 4;
```
@alexcrichton here is another place where I'd like to simplify things, is there a better way of obtaining a pointer as u32 number?

alexcrichton replied:
I actually think it's best to simply cahnge `wait_async` to take `&AtomicI32`, and then you can do different things internally there.

ibaryshnikov commented:
@alexcrichton thank you for a great review! I fixed most of the places, but still have some questions, like usage of `AtomicI32`, or simplifying `Waker` struct. Btw, it looks working, and after increasing default timeout from **0** to 1**0ms** the performance looks fine. So it's probably a time to start making changes to raytrace_parallel example itself?

alexcrichton commented on crates/futures/src/lib.rs:
```diff
@@ -101,12 +101,21 @@
+#![feature(stdsimd)]
 #![deny(missing_docs)]
 
 #[cfg(feature = "futures_0_3")]
```
While you're at it it's probably good to remove this if `"atomics"` is enabled (since it's not compatible).

The `cfg-if` crate may be able to help clean up the `#[cfg]` here a bit too.

alexcrichton commented on crates/futures/src/lib.rs:
```diff
@@ -101,12 +101,21 @@
+/// Polyfill for `Atomics.waitAsync` function
+mod polyfill;
+
 use std::cell::{Cell, RefCell};
```
Oh so actually the existing implementation in this file should be removed when `atomics` is enabled. With atomics the current implementation is unsound, but the `atomics` module contains the same interface anyway so the idea is that we'd just swap out for different implementations depending on `atomics` or not.

alexcrichton commented on crates/futures/src/polyfill.rs:
```diff
@@ -0,0 +1,208 @@
+        let encoded: String = encode_uri_component(HELPER_CODE).into();
+        initialization_string.push_str(&encoded);
+
+        return Rc::new(Worker::new(&initialization_string).expect("Should create a Worker"));
```
Sorry this may have gotten lost in the previous review comments, but I think that the `Rc` here shouldn't be necessary because a `Worker` can be natively cloned?

alexcrichton commented on crates/futures/src/polyfill.rs:
```diff
@@ -0,0 +1,208 @@
+    })
+}
+
+fn free_helper(helper: &Rc<Worker>) {
```
This might be a bit more appropriate to take `helper` by value rather than behind a reference

alexcrichton commented on crates/futures/src/polyfill.rs:
```diff
@@ -0,0 +1,208 @@
+}
+
+pub fn wait_async(index: u32, value: i32) -> Result<Promise, JsValue> {
+    wait_async_with_timeout(index, value, DEFAULT_TIMEOUT)
```
I think that instead of a `DEFAULT_TIMEOUT` you can pass `-1` here which indicates "wait forever"

Or at least I think that's the intented API of  `Atomics.waitAsync`

ibaryshnikov replied:
In the docs it says that it defaults to `Infinity`, so I just made it as default. I didn't find any info about `-1` in the docs, and desided not to use it

alexcrichton commented on crates/futures/src/polyfill.rs:
```diff
@@ -0,0 +1,208 @@
+    });
+}
+
+pub fn wait_async(index: u32, value: i32) -> Result<Promise, JsValue> {
```
FWIW I'd still recommend changing `index: u32` to `ptr: &AtomicI32`

alexcrichton commented on crates/futures/src/polyfill.rs:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    // Optimization, avoid the helper thread in this common case.
+    if Atomics::load(&indexed_array, index)? != value {
```
If `index` is changed to `&AtomicI32`, you don't need to call a JS API here and can instead simply do `ptr.load(SeqCst)`

alexcrichton commented on crates/futures/src/polyfill.rs:
```diff
@@ -0,0 +1,208 @@
+
+    let indexed_array = Int32Array::new(&memory_buffer);
+
+    if !indexed_array.buffer().has_type::<SharedArrayBuffer>() {
```
This I think is fine to elide since it's always going to be true :)

alexcrichton commented on crates/futures/src/polyfill.rs:
```diff
@@ -0,0 +1,208 @@
+                as Box<dyn FnMut(MessageEvent)>);
+            helper.set_onmessage(Some(onmessage_callback.as_ref().unchecked_ref()));
+
+            // onmessage_callback.forget();
```
The instance of `Closure` here needs to be persisted with the `Worker` to ensure it's not deallocated on the Rust side of things, but this is basically stored with `helper_ref` I think

ibaryshnikov replied:
Do you mean to create something like
```rust
struct WorkerData {
  worker: Worker,
  closure: Closure,
}
```
?

ibaryshnikov commented:
@alexcrichton I fixed most of the places, but still have some where I'm not sure, like what to do with closures, and the contents of the `Waker` struct, which is currently
```rust
struct Waker {
    value: AtomicI32,
    notified: AtomicBool,
}
```

alexcrichton commented on crates/futures/src/atomics.rs:
```diff
@@ -0,0 +1,326 @@
+        // helps avoid blowing the stack by accident.
+        let promise =
+            crate::polyfill::wait_async(&package.waker.value).expect("Should create a Promise");
+        let closure = Closure::once(Box::new(move |_| {
```
FWIW I think [`once_into_js`](https://docs.rs/wasm-bindgen/0.2.47/wasm_bindgen/closure/struct.Closure.html#method.once_into_js) is what you want here

ibaryshnikov replied:
I don't get it, once_into_js returns JsValue, and promise.then expects Closure. Am I missing something?

alexcrichton replied:
Ah this needed a redefinition of a few functions, but I've handled that and pushed a few commits.

ibaryshnikov commented:
Updated to the latest master (I hope I didn't miss anything). Removed `notified` field, there's only `value` now inside `Waker`.
Can we summarize, what's left to do in this pr? I'm kind of lost in the comments.

alexcrichton commented:
Ok thanks for the update @ibaryshnikov! 

I've added a few more commits which cleaned up a few things (good chunks of the old `future_to_promise` implementation aren't necessary when using `waitAsync`) and reorganized some files a bit. I also hooked up the parallel raytrace example and can confirm it's all plumbed together and at least rudimentarily working.

If CI is green and you're good with this I think it's good to merge!

ibaryshnikov commented:
@alexcrichton I'm good with this. Thanks for moving this forvard! And thanks for your comments, I've learned a lot from this pr!)

alexcrichton commented:
Ok let's merge then, thanks so much again @ibaryshnikov!
